torchtitan/clip_grad_nrom.py

```diff
@@ -0,0 +1,137 @@
+            If ``None``, use the foreach implementation for CUDA and CPU native tensors and silently
+            fall back to the slow implementation for other device types.
+            Default: ``None``
+        pp_mesh: pipeline parallel device mesh. If not None, will reduce gradient norm across PP stages.
+
+    Returns:
+        Total norm of the parameter gradients (viewed as a single vector).
+
+    """
+    if pp_mesh is None:
+        # fallback to the original implementation
+        return torch.nn.utils.clip_grad_norm_(
+            parameters,
+            max_norm=max_norm,
+            norm_type=norm_type,
+            error_if_nonfinite=error_if_nonfinite,
+            foreach=foreach,
+        )
+
+    """
+    Gradient norm clipping requires computing the gradient norm over the entire model.
+    `torch.nn.utils.clip_grad_norm_` only computes gradient norm along DP/FSDP/TP dimensions.
+    We need to manually reduce the gradient norm across PP stages.
+    See https://github.com/pytorch/torchtitan/issues/596 for details.
+    """
+    if isinstance(parameters, torch.Tensor):
+        parameters = [parameters]
+    grads = [p.grad for p in parameters if p.grad is not None]
+    max_norm = float(max_norm)
+    norm_type = float(norm_type)
+    if len(grads) == 0:
+        return torch.tensor(0.0)
+    first_device = grads[0].device
+    grouped_grads: Dict[
+        Tuple[torch.device, torch.dtype], Tuple[List[List[torch.Tensor]], List[int]]
+    ] = _group_tensors_by_device_and_dtype(
+        [grads]
+    )  # type: ignore[assignment]
+
+    norms: List[torch.Tensor] = []
+    for (device, _), ([device_grads], _) in grouped_grads.items():  # type: ignore[assignment]
+        if (foreach is None and _has_foreach_support(device_grads, device)) or (
+            foreach and _device_has_foreach_support(device)
+        ):
+            norms.extend(torch._foreach_norm(device_grads, norm_type))
+        elif foreach:
+            raise RuntimeError(
+                f"foreach=True was passed, but can't use the foreach API on {device.type} tensors"
+            )
+        else:
+            norms.extend([torch.linalg.vector_norm(g, norm_type) for g in device_grads])
+
+    total_norm = torch.linalg.vector_norm(
+        torch.stack([norm.to(first_device) for norm in norms]), norm_type
+    )
+
+    # ----- start modified from torch.nn.utils.clip_grad_norm_ -----
+    if isinstance(total_norm, DTensor):
+        # if total_norm is a DTensor, the placements must be `torch.distributed._tensor.ops.math_ops._NormPartial`
+        # we can simply reduce the DTensor to get the total norm in this tensor's process group
+        # and then convert it to a local tensor
+        total_norm = total_norm.redistribute(
+            placements=[Replicate()] * total_norm.device_mesh.ndim
+        ).to_local()
+
+    total_norm **= norm_type
+    dist.all_reduce(total_norm, op=dist.ReduceOp.SUM, group=pp_mesh.get_group())
```

Comment:
I don't like that we have to copy/modify the whole clip_grad_norm, and I think I see the reason why we do it and might have an idea for avoiding it.

IIUC the issue is that clip_grad_norm bundles 2 actions into one API.
1. compute total norm
2. apply clipping based on total norm

would it be cleaner if we provided unbundled APIs at the torch layer for these 2?  then in the PP case we can just call the first helper, do the PP reduce, then call the second helper.

Reply:
exactly what we need!

Reply:
It would be great if `torch.nn.utils.clip_grad_norm_` can be split into 2 APIs

Reply:
@awgu should we create a github issue/feature request under pytorch main repo?

Reply:
Yes let us do that. I can ping people internally after we make the issue too.

Reply:
@awgu issue in pytorch main repo created. See https://github.com/pytorch/pytorch/issues/139467

Reply:
Hi @zijian-hu, thank you for your changes so far. It looks like https://github.com/pytorch/pytorch/pull/139662 is getting checked in, could you update your PR to use these APIs? Thanks so much!

Reply:
@H-Huang Just saw the pytorch issue has been resolved. Will update this PR ASAP

Reply:
@H-Huang Updated with the newest PyTorch clip norm APIs